File: day19/day19.py
import cProfile
_input = open("input", 'r').readlines()

def day19():
	opcodes = {
		"addr": lambda A, B, reg: reg[A] + reg[B],
		"addi": lambda A, B, reg: reg[A] + B,
		"mulr": lambda A, B, reg: reg[A] * reg[B],
		"muli": lambda A, B, reg: reg[A] * B,
		"banr": lambda A, B, reg: reg[A] & reg[B],
		"bani": lambda A, B, reg: reg[A] & B,
		"borr": lambda A, B, reg: reg[A] | reg[B],
		"bori": lambda A, B, reg: reg[A] | B,
		"setr": lambda A, B, reg: reg[A],
		"seti": lambda A, B, reg: A,
		"gtir": lambda A, B, reg: int(A > reg[B]),
		"gtri": lambda A, B, reg: int(reg[A] > B),
		"gtrr": lambda A, B, reg: int(reg[A] > reg[B]),
		"eqir": lambda A, B, reg: int(A == reg[B]),
		"eqri": lambda A, B, reg: int(reg[A] == B),
		"eqrr": lambda A, B, reg: int(reg[A] == reg[B]),
	}

	ipReg = int(_input[0][4])
	instructions = [line[:-1].split(' ') for line in _input[1:]]

	registers = [0,0,0,0,0,0]
	ip = 0
	while ip < len(instructions):
		registers[ipReg] = ip
		inst = instructions[ip]
		op = inst[0]
		A = int(inst[1])
		B = int(inst[2])
		C = int(inst[3])
		registers[C] = opcodes[op](A,B,registers)

		ip = registers[ipReg] + 1

	print(registers)

cProfile.run("day19()")

File: day19/c/day19.c
#include <stdio.h>
#include <stdlib.h>

int reg[6] = {1,0,0,0,0,0};

int addr(int A, int B) {
	return reg[A] + reg[B];
}
int addi(int A, int B) {
	return reg[A] + B;
}
int mulr(int A, int B) {
	return reg[A] * reg[B];
}
int muli(int A, int B) {
	return reg[A] * B;
}
int setr(int A, int B) {
	return reg[A];
}
int seti(int A, int B) {
	return A;
}
int gtrr(int A, int B) {
	return (int)(reg[A] > reg[B]);
}
int eqrr(int A, int B) {
	return (int)(reg[A] == reg[B]);
}

int (*ptf[8])(int A, int B) = {addr,addi,mulr,muli,setr,seti,gtrr,eqrr};

void main() {
	int **pgm = (int**)malloc(36*sizeof(int*));
	for(int i = 0; i < 36; i++) {
		pgm[i] = malloc(4*sizeof(int));
	}

	FILE *input = fopen("A:\\Documents\\aoc-2018\\day19\\c\\input", "r");
	if (input == NULL) {
		exit(EXIT_FAILURE);
	}

	for (int i = 0; i < 36; i++) {
		for (int j = 0; j < 4; j++) {
			if (!fscanf(input, "%ld", &pgm[i][j]))
				break;
		}
	}
	fclose(input);

	int ipReg = 5;
	int ip = 0;

	while(ip < 36) {
		reg[ipReg] = ip;
		int *inst = pgm[ip];
		reg[inst[3]] = (*ptf[inst[0]])(inst[1], inst[2]);

		ip = reg[ipReg] + 1;
	}

	for(int i = 0; i<6; i++) {
		printf("%d ", reg[i]);
	}
	printf("\n");

	for (int i = 0; i < 36; i++) {
		free(pgm[i]);
	}
	free(pgm);
}